File: web-greeter/resources/js/ThemeUtils.js
/*
 * GreeterUtils.js
 *
 * Copyright © 2017 Antergos Developers <dev@antergos.com>
 *
 * This file is part of lightdm-webkit2-greeter.
 *
 * lightdm-webkit2-greeter is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * lightdm-webkit2-greeter is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * The following additional terms are in effect as per Section 7 of the license:
 *
 * The preservation of all legal notices and author attributions in
 * the material or in the Appropriate Legal Notices displayed
 * by works containing it is required.
 *
 * You should have received a copy of the GNU General Public License
 * along with lightdm-webkit2-greeter; If not, see <http://www.gnu.org/licenses/>.
 */


let localized_invalid_date = null,
	time_language = null,
	time_format = null,
	allowed_dirs = null,
	_ThemeUtils = null;


function _set_allowed_dirs() {
	allowed_dirs = {
		themes_dir: lightdm.themes_dir,
		backgrounds_dir: greeter_config.branding.background_images_dir,
		lightdm_data_dir: lightdm.shared_data_dir,
		tmpdir: '/tmp',
	};
}



/**
 * Provides various utility methods for use in greeter themes. The greeter will automatically
 * create an instance of this class when it starts. The instance can be accessed
 * with the global variable: [`theme_utils`](#dl-window-theme_utils).
 *
 * @memberOf LightDM
 */
class ThemeUtils {

	constructor( instance ) {
		if ( null !== _ThemeUtils ) {
			return _ThemeUtils;
		}

		moment.locale( window.navigator.languages );

		localized_invalid_date = moment('today', '!@#');
		_ThemeUtils = instance;
	}

	/**
	 * Binds `this` to class, `context`, for all of the class's methods.
	 *
	 * @arg {object} context An ES6 class instance with at least one method.
	 *
	 * @return {object} `context` with `this` bound to it for all of its methods.
	 */
	bind_this( context ) {
		let excluded_methods = ['constructor'];

		function not_excluded( _method, _context ) {
			let is_excluded = excluded_methods.findIndex( excluded_method => _method === excluded_method ) > -1,
				is_method = 'function' === typeof _context[_method];

			return is_method && !is_excluded;
		}

		for ( let obj = context; obj; obj = Object.getPrototypeOf( obj ) ) {
			// Stop once we have traveled all the way up the inheritance chain
			if ( 'Object' === obj.constructor.name ) {
				break;
			}

			for ( let method of Object.getOwnPropertyNames( obj ) ) {
				if ( not_excluded( method, context ) ) {
					context[method] = context[method].bind( context );
				}
			}
		}

		return context;
	}


	/**
	 * Returns the contents of directory found at `path` provided that the (normalized) `path`
	 * meets at least one of the following conditions:
	 *   * Is located within the greeter themes' root directory.
	 *   * Has been explicitly allowed in the greeter's config file.
	 *   * Is located within the greeter's shared data directory (`/var/lib/lightdm-data`).
	 *   * Is located in `/tmp`.
	 *
	 * @param {string}              path        The abs path to desired directory.
	 * @param {boolean}             only_images Include only images in the results. Default `true`.
	 * @param {function(string[])}  callback    Callback function to be called with the result.
	 */
	dirlist( path, only_images = true, callback ) {
		let allowed = false;

		if ( '' === path || 'string' !== typeof path ) {
			console.log('[ERROR] theme_utils.dirlist(): path must be a non-empty string!');
			return callback([]);

		} else if ( null !== path.match(/^[^/].+/) ) {
			console.log('[ERROR] theme_utils.dirlist(): path must be absolute!');
			return callback([]);
		}

		if ( null !== path.match(/\/\.+(?=\/)/) ) {
			// No special directory names allowed (eg ../../)
			path = path.replace(/\/\.+(?=\/)/g, '' );
		}

		if ( null === allowed_dirs ) {
			_set_allowed_dirs();
		}

		if ( ! Object.keys( allowed_dirs ).some( dir => path.startsWith( allowed_dirs[dir] ) ) ) {
			console.log(`[ERROR] theme_utils.dirlist(): path is not allowed: ${path}`);
			return callback([]);
		}

		try {
			return _ThemeUtils.dirlist( path, only_images, callback );

		} catch( err ) {
			console.log( `[ERROR] theme_utils.dirlist(): ${err}` );
			return callback([]);
		}
	}

	/**
	 * Escape HTML entities in a string.
	 *
	 * @param {string} text The text to be escaped.
	 *
	 * @returns {string}
	 */
	esc_html( text ) {
		return this.txt2html( text );
	}


	/**
	 * Get the current time in a localized format. Time format and language are auto-detected
	 * by default, but can be set manually in the greeter config file.
	 *   * `language` defaults to the system's language, but can be set manually in the config file.
	 *   * When `time_format` config file option has a valid value, time will be formatted
	 *     according to that value.
	 *   * When `time_format` does not have a valid value, the time format will be `LT`
	 *     which is `1:00 PM` or `13:00` depending on the system's locale.
	 *
	 * @return {string} The current localized time.
	 */
	get_current_localized_time() {
		if ( null === time_language ) {
			let config = greeter_config.greeter,
				manual_language = ( '' !== config.time_language && 'auto' !== config.time_language ),
				manual_time_format = ( '' !== config.time_format && 'auto' !== config.time_format );

			time_language =  manual_language ? config.time_language : window.navigator.language;
			time_format = manual_time_format ? config.time_format : 'LT';

			if ( manual_language ) {
				moment.locale( time_language );
			}
		}

		let local_time = moment().format( time_format );

		if ( local_time === localized_invalid_date ) {
			local_time = moment().format( 'LT' );
		}

		return local_time;
	}


	/**
	 * Use {@link window.theme_utils.esc_html()} instead.
	 *
	 * @deprecated
	 */
	txt2html( text ) {
		try {
			return _ThemeUtils.esc_html( text );

		} catch( err ) {
			console.log( `[ERROR] theme_utils.dirlist(): ${err}` );
			return text;
		}
	}
}

File: web-greeter/bridge/ThemeUtils.py
# -*- coding: utf-8 -*-
#
#  ThemeUtils.py
#
#  Copyright © 2017 Antergos
#
#  This file is part of Web Greeter.
#
#  Web Greeter is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 3 of the License, or
#  (at your option) any later version.
#
#  Web Greeter is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  The following additional terms are in effect as per Section 7 of the license:
#
#  The preservation of all legal notices and author attributions in
#  the material or in the Appropriate Legal Notices displayed
#  by works containing it is required.
#
#  You should have received a copy of the GNU General Public License
#  along with Web Greeter; If not, see <http://www.gnu.org/licenses/>.

# Standard Lib
import os
from glob import glob
import tempfile

# 3rd-Party Libs
from whither.bridge import (
    BridgeObject,
    bridge,
    Variant,
)


class ThemeUtils(BridgeObject):

    def __init__(self, greeter, config, *args, **kwargs):
        super().__init__(name='ThemeUtils', *args, **kwargs)

        self._config = config
        self._greeter = greeter

        self._allowed_dirs = (
            self._config.themes_dir,
            self._config.branding.background_images_dir,
            self._greeter.shared_data_directory,
            tempfile.gettempdir(),
        )

    @bridge.method(str, bool, result=Variant)
    def dirlist(self, dir_path, only_images=True):
        if not dir_path or not isinstance(dir_path, str) or '/' == dir_path:
            return []

        dir_path = os.path.realpath(os.path.normpath(dir_path))

        if not os.path.isabs(dir_path) or not os.path.isdir(dir_path):
            return []

        allowed = False

        for allowed_dir in self._allowed_dirs:
            if dir_path.startswith(allowed_dir):
                allowed = True
                break

        if not allowed:
            return []

        if only_images:
            file_types = ('jpg', 'jpeg', 'png', 'gif', 'bmp', 'webp')
            result = [
                glob('{0}/**/*.{1}'.format(dir_path, ftype), recursive=True)
                for ftype in file_types
            ]
            result = [image for image_list in result for image in image_list]

        else:
            result = [os.path.join(dir_path, f) for f in os.listdir(dir_path)]

        return result
